File: Leetcode-Problems/543 - DiameterofBinaryTree.js
/**
 * Given the root of a binary tree, return the length of the diameter of the tree.
 * The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.
 * The length of a path between two nodes is represented by the number of edges between them.
 * 
 * Time Complexity: O(N), where N is the number of nodes in the tree.
 * - Each node is visited once in a Depth-First Search (DFS).
 *
 * Space Complexity: O(H), where H is the height of the tree.
 * - In the worst case (skewed tree), the recursion stack can go up to O(N).
 * - In the best case (balanced tree), the recursion stack is O(log N).
 */
const diameterOfBinaryTree = (root) => {
    // Edge case: If the tree is empty, return diameter as 0.
    if (!root) return 0;

    let max = 0; // Variable to track the maximum diameter found.

    /**
     * Performs a Depth-First Search (DFS) to compute the height of each subtree.
     * Updates the maximum diameter encountered during traversal.
     */
    const dfs = (node) => {
        if (!node) return 0; // Base case: null nodes contribute 0 height.

        // Recursively find the height of the left and right subtrees.
        let left = dfs(node.left);
        let right = dfs(node.right);

        // Update max diameter: Sum of left and right subtree heights at this node.
        max = Math.max(left + right, max);

        // Return the height of the current node's subtree (1 + max child height).
        return Math.max(left, right) + 1;
    };

    dfs(root); // Start DFS from the root node.

    return max; // Return the maximum diameter found.
};

File: Platforms(LeetCode, CF, CC)/LeetCode/543-DiameterOfBinaryTree.cpp
// ===============================================
// Problem: Diameter of Binary Tree
// 
// Given a binary tree, find the length of the longest path between any two nodes.
// The path may or may not pass through the root.
//
// Time Complexity: O(n)
//   - Each node is visited exactly once.
//
// Space Complexity: O(h)
//   - h = height of tree (stack space for recursion)
//   - Worst case: O(n) skewed tree
//   - Best case: O(log n) balanced tree
// ===============================================

#include <bits/stdc++.h>
using namespace std;

// -----------------------------------------------
// Definition for a binary tree node
// -----------------------------------------------
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};

class Solution
{
    // ------------------------------------------------------------
    // dfs(root, ans):
    // Computes the height of subtree rooted at `root`.
    // Also updates `ans` to store the largest diameter seen so far.
    //
    // leftDepth  = height of left subtree
    // rightDept  = height of right subtree
    //
    // Diameter through current node = leftDepth + rightDepth + 1 (number of nodes)
    //
    // Returns: height of the current subtree
    // ------------------------------------------------------------
    int dfs(TreeNode *root, int &ans)
    {
        if(root == nullptr) return 0; // Base case: empty node has height 0

        // Recursively compute heights of left and right subtrees
        int leftDepth = dfs(root->left, ans);
        int rightDept = dfs(root->right, ans);

        // Update the maximum diameter seen so far
        ans = max(ans, leftDepth + rightDept + 1);

        // Return height of current subtree (count this node)
        return max(leftDepth, rightDept) + 1;
    }

public:
    // ------------------------------------------------------------
    // diameterOfBinaryTree(root):
    // Wrapper function that calls dfs to compute the diameter.
    // Returns diameter in terms of edges (LeetCode convention)
    // ------------------------------------------------------------
    int diameterOfBinaryTree(TreeNode *root)
    {
        int ans = 0;
        dfs(root, ans);
        return ans - 1; // Convert from node count to edge count
    }
};